crates/spirv-std/src/derivative.rs

```diff
@@ -0,0 +1,98 @@
+//! The [`Derivative`] trait for getting derivatives and handling derivative
+//! operations in SPIR-V.
+
+/// Represents a type that can represent the derivation of an operation.
+pub trait Derivative {
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// X coordinate. Returns the same result as either
+    /// [`Self::ddx_fine`] or [`Self::ddx_coarse`], selection of which one is
+    /// dependent on external factors.
+    fn ddx(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// X coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment and its immediate neighbor(s).
+    fn ddx_fine(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// X coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment’s neighbors, and possibly, but not necessarily,
+    /// includes the value of `Self` for the current fragment. That is, over a
+    /// given area, the implementation can compute X derivatives in fewer
+    /// unique locations than would be allowed by [`Self::ddx_fine`].
+    fn ddx_coarse(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// Y coordinate. Returns the same result as either [`Self::ddy_fine`] or
+    /// [`Self::ddy_coarse`], selection of which one is dependent on
+    /// external factors.
+    fn ddy(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// Y coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment and its immediate neighbor(s).
+    fn ddy_fine(self) -> Self;
+    /// Returns the partial derivative of `Self` with respect to the window's
+    /// Y coordinate. Uses local differencing based on the value of `Self` for
+    /// the current fragment’s neighbors, and possibly, but not necessarily,
+    /// includes the value of `Self` for the current fragment. That is, over a
+    /// given area, the implementation can compute Y derivatives in fewer
+    /// unique locations than would be allowed by [`Derivative::ddy_fine`].
+    fn ddy_coarse(self) -> Self;
+    /// Returns the sum of [`Self::ddx`] and [`Self::ddy`] as a
```

Comment:
nit: this is `abs(ddx())+abs(ddy())`, not `ddx()+ddy()` (I would imagine most people reading these docs would already know that, but idk)

Reply:
> (I would imagine most people reading these docs would already know that, but idk)

No you're right, it should mention that. I just glossed over the words *absolute values* in the spec. Best to make no assumptions about the readers.